### ISSUE TITLE
英特尔 oneAPI - Deep Neural Network Library (oneDNN) 的初识与示例

### ISSUE THREAD

**Issue by ninerui** [ninerui/ninerui.github.io#12]
# 介绍
## 什么是oneAPI
oneAPI 是一种开放的、跨行业、基于标准的、统一的、多架构、多供应商的编程模型，可跨加速器架构提供通用的开发人员体验，从而实现更快的应用程序性能、更高的生产力和更大的创新。oneAPI计划鼓励在整个生态系统中就oneAPI规范和兼容的oneAPI实施进行协作。

## 什么是oneDNN
oneAPI 深度神经网络库 (oneDNN) 是一个开源的跨平台高性能库，其中包含用于深度学习应用程序的基本模块。oneDNN 是 [oneAPI](https://oneapi.io/) 的一部分。该库针对英特尔(R) 架构处理器、英特尔处理器显卡和 Xe 架构显卡进行了优化。
oneDNN 支持基于以下体系结构的平台：
- Arm* 64-bit Architecture (AArch64)
- NVIDIA* GPU
- OpenPOWER* Power ISA (PPC64)
- IBMz* (s390x)
- RISC-V

oneDNN 编程模型包括一个或多个基元来处理一个或多个内存对象中的数据。执行是在流的上下文中的引擎上执行的。下图简要介绍了这些实体之间的关系。
![image](https://user-images.githubusercontent.com/43716635/167803757-736b357b-d1e2-4262-86f3-87d3abfb6810.png)
## 基元
oneDNN是围绕基元（dnnl::primitive）的概念构建的。基元是封装特定计算（如正向卷积、后向 LSTM 计算或数据转换操作）的对象。此外，使用基元属性（dnnl::primitive_attr），某些基元可以表示更复杂的融合计算，如前向卷积后的ReLU。
## 引擎
引擎（dnnl::engine）是计算设备的抽象：CPU、系统中的特定GPU卡等。创建大多数基元是为了在一个特定引擎上执行计算。
## 流
流（dnnl::[stream]）封装了绑定到特定引擎执行的上下文。例如，它们可以对应于 OpenCL 命令队列。
## 内存对象
内存对象 （[dnnl::memory]） 将句柄封装到在特定引擎、张量维度、数据类型和内存格式上分配的内存中 ， 张量索引映射到线性内存空间中偏移量的方式。内存对象在执行期间传递给基元。

# 编译安装
## 下载源代码
```git clone https://github.com/oneapi-src/oneDNN.git```
## oneAPI DPC++ 编译器
- 为 oneAPI DPC++ 编译器设置环境。对于安装到默认位置的英特尔 oneAPI 基础工具包发行版，您可以使用以下脚本执行此操作：setvars.sh
```source /opt/intel/oneapi/setvars.sh```
- 配置 CMake 并生成生成文件
```
mkdir -p build
cd build

export CC=icx
export CXX=icpx

cmake .. \
          -DDNNL_CPU_RUNTIME=DPCPP
          -DDNNL_GPU_RUNTIME=DPCPP
          <extra build options>
```
- 构建库
```make -j```
- 验证
```ctest```
![image](https://user-images.githubusercontent.com/43716635/168010721-ce79e989-c530-42af-bf6d-d870b0546516.png)
- 构建文档
```cmake --build . --target doc```
- 安装库
```cmake --build . --target install```

# 代码示例
```
#include <cmath>
#include <numeric>
#include <stdexcept>
#include <vector>

#include "oneapi/dnnl/dnnl.hpp"
#include "oneapi/dnnl/dnnl_debug.h"

#include "example_utils.hpp"

using namespace dnnl;

void getting_started_tutorial(engine::kind engine_kind) {
    // 初始化引擎
    engine eng(engine_kind, 0);

    // 初始化流
    stream engine_stream(eng);

    // 创建数据
    const int N = 1, H = 13, W = 13, C = 3;

    // 计算每个维度的物理步幅
    const int stride_N = H * W * C;
    const int stride_H = W * C;
    const int stride_W = C;
    const int stride_C = 1;

    // 将逻辑索引映射到物理偏移量
    auto offset = [=](int n, int h, int w, int c) {
        return n * stride_N + h * stride_H + w * stride_W + c * stride_C;
    };

    // 图片大小
    const int image_size = N * H * W * C;

    // 为图像分配缓存
    std::vector<float> image(image_size);

    // 为图像初始化一些值
    for (int n = 0; n < N; ++n)
        for (int h = 0; h < H; ++h)
            for (int w = 0; w < W; ++w)
                for (int c = 0; c < C; ++c) {
                    int off = offset(
                            n, h, w, c); // Get the physical offset of a pixel
                    image[off] = -std::cos(off / 10.f);
                }

    auto src_md = memory::desc(
            {N, C, H, W}, // 顺序由基元决定
            memory::data_type::f32, // 张量的数据类型
            memory::format_tag::nhwc // 内存格式，这种情况下为NHWC
    );

    auto alt_src_md = memory::desc(
            {N, C, H, W}, // 顺序由基元决定
            memory::data_type::f32, // 张量的数据类型
            {stride_N, stride_C, stride_H, stride_W}
    );

    // 检查内存描述符是否一致
    if (src_md != alt_src_md)
        throw std::logic_error("Memory descriptor initialization mismatch.");

    auto src_mem = memory(src_md, eng);
    write_to_dnnl_memory(image.data(), src_mem);

    // 为 dst_mem 分配缓存
    auto dst_mem = memory(src_md, eng);

    auto relu_d = eltwise_forward::desc(
            prop_kind::forward_inference, algorithm::eltwise_relu,
            src_md,
            0.f,
            0.f
    );

    auto relu_pd = eltwise_forward::primitive_desc(
            relu_d, eng
    );

    auto relu = eltwise_forward(relu_pd);

    relu.execute(engine_stream, {
            {DNNL_ARG_SRC, src_mem},
            {DNNL_ARG_DST, dst_mem},
    });

    engine_stream.wait();

    std::vector<float> relu_image(image_size);
    read_from_dnnl_memory(relu_image.data(), dst_mem);
    for (int n = 0; n < N; ++n)
        for (int h = 0; h < H; ++h)
            for (int w = 0; w < W; ++w)
                for (int c = 0; c < C; ++c) {
                    int off = offset(
                            n, h, w, c); // get the physical offset of a pixel
                    float expected = image[off] < 0
                                     ? 0.f
                                     : image[off]; // expected value
                    if (relu_image[off] != expected) {
                        std::cout << "At index(" << n << ", " << c << ", " << h
                                  << ", " << w << ") expect " << expected
                                  << " but got " << relu_image[off]
                                  << std::endl;
                        throw std::logic_error("Accuracy check failed.");
                    }
                }
}

int main(int argc, char **argv) {
    int exit_code = 0;

    engine::kind engine_kind = parse_engine_kind(argc, argv);
    try {
        getting_started_tutorial(engine_kind);
    } catch (dnnl::error &e) {
        std::cout << "oneDNN error caught: " << std::endl
                  << "\tStatus: " << dnnl_status2str(e.status) << std::endl
                  << "\tMessage: " << e.what() << std::endl;
        exit_code = 1;
    } catch (std::string &e) {
        std::cout << "Error in the example: " << e << "." << std::endl;
        exit_code = 2;
    }

    std::cout << "Example " << (exit_code ? "failed" : "passed") << " on "
              << engine_kind2str_upper(engine_kind) << "." << std::endl;
    return exit_code;
}
```
此代码示例演示了 oneDNN 编程的基础知识，其中代码流程是：
- 创建引擎和流用来执行基元
- 执行数据准备
- 将数据包装到 oneDNN 内存对象中
- 创建 ReLU 基元
ReLu又称修正线性单元，是一种深度学习中常用的激活函数（activation function）。
![image](https://user-images.githubusercontent.com/43716635/168228196-15c69199-e12f-4ef8-a443-f6a3d32fc478.png)
![image](https://user-images.githubusercontent.com/43716635/168227789-ae24e3c8-87c1-4162-a43d-7cffce65300e.png)
- 执行 ReLU 基元
- 获取结果并进行验证

# 其它示例
- [Convolution](https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html)
- [Inner Product](https://oneapi-src.github.io/oneDNN/dev_guide_inner_product.html)
- [Matrix Multiplication](https://oneapi-src.github.io/oneDNN/dev_guide_matmul.html)
- [RNN](https://oneapi-src.github.io/oneDNN/dev_guide_rnn.html)
- [Batch Normalization](https://oneapi-src.github.io/oneDNN/dev_guide_batch_normalization.html)
- [Binary](https://oneapi-src.github.io/oneDNN/dev_guide_binary.html)
- [Concat](https://oneapi-src.github.io/oneDNN/dev_guide_concat.html)
- [Eltwise](https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html)
- [Layer Normalization](https://oneapi-src.github.io/oneDNN/dev_guide_layer_normalization.html)
- [Local Response Normalization (LRN)](https://oneapi-src.github.io/oneDNN/dev_guide_lrn.html)
- [LogSoftmax](https://oneapi-src.github.io/oneDNN/dev_guide_logsoftmax.html)
- [Pooling](https://oneapi-src.github.io/oneDNN/dev_guide_pooling.html)
- [PReLU](https://oneapi-src.github.io/oneDNN/dev_guide_prelu.html)
- [Resampling](https://oneapi-src.github.io/oneDNN/dev_guide_resampling.html)
- [Shuffle](https://oneapi-src.github.io/oneDNN/dev_guide_shuffle.html)
- [Softmax](https://oneapi-src.github.io/oneDNN/dev_guide_softmax.html)
- [Sum](https://oneapi-src.github.io/oneDNN/dev_guide_sum.html)
- [Reorder](https://oneapi-src.github.io/oneDNN/dev_guide_reorder.html)
- [Reduction](https://oneapi-src.github.io/oneDNN/dev_guide_reduction.html)

oneDnn 官方支持的基元还算比较丰富，包含了绝大部分深度学习模块

# 二分类交叉熵损失函数
二分类任务中，因为只有正例和负例，且两者的概率和是1，所以损失函数定义简化如下：
![image](https://user-images.githubusercontent.com/43716635/168256077-0b4c5f49-6db9-4f9c-be33-4be13585b860.png)
其中 ![yy](https://latex.codecogs.com/svg.image?\hat{y}) 是模型预测样本是正例的概率， ![y](https://latex.codecogs.com/svg.image?y) 是样本标签，如果样本属于正例，取值为1，否则取值为0。
使用oneDNN代码实现示例
```
#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

#include "example_utils.hpp"
#include "oneapi/dnnl/dnnl.hpp"

using namespace dnnl;

using tag = memory::format_tag;
using dt = memory::data_type;

// 二分类交叉熵损失函数
void binary_cross_entropy_loss(
        dnnl::engine::kind engine_kind,
        dnnl::memory y_pred,
        dnnl::memory y_true) {

    dnnl::engine engine(engine_kind, 0);
    dnnl::stream engine_stream(engine);

    memory::dims dim_nc = {y_pred.get_desc().dims()[0], 1};
    memory::dims dim_cn = {1, y_pred.get_desc().dims()[0]};
    memory::dims dim_scalar = {1, 1};

    // 创建向量
    memory::desc vector_md = memory::desc(dim_nc, dt::f32, tag::nc);
    // 创建向量转置
    memory::desc vector_transpose_md = memory::desc(dim_cn, dt::f32, tag::cn);
    // 创建标量
    memory::desc scalar_md = memory::desc(dim_scalar, dt::f32, tag::nc);

    // 防止y_pred报错添加限制
    float y_pred_upper = 1 - 1e-7;
    float y_pred_lower = 1e-7;
    auto y_pred_clip = dnnl::memory(vector_transpose_md, engine);
    auto clip_desc = eltwise_forward::desc(
            prop_kind::forward_training,
            algorithm::eltwise_clip,
            vector_md,
            y_pred_lower,
            y_pred_upper);
    auto clip_pd = dnnl::eltwise_forward::primitive_desc(clip_desc, engine);

    // 对y_pred执行log
    auto y_pred_log = dnnl::memory(vector_transpose_md, engine);
    auto log_desc = eltwise_forward::desc(
            prop_kind::forward_training,
            algorithm::eltwise_log,
            vector_transpose_md,
            0.f,
            0.f);
    auto log_pd = eltwise_forward::primitive_desc(log_desc, engine);

    // 在 y_pred 执行 alpha = -1, beta = 1
    auto y_pred_inv = dnnl::memory(vector_transpose_md, engine);
    auto inv_desc = eltwise_forward::desc(
            prop_kind::forward_training,
            algorithm::eltwise_linear,
            vector_transpose_md,
            -1.f,
            1.f);
    auto inv_pd = eltwise_forward::primitive_desc(inv_desc, engine);

    // 计算log(1-y_pred)
    auto y_pred_inv_log = dnnl::memory(vector_transpose_md, engine);

    // 在 y_true 执行 alpha = -1, beta = 1
    auto y_true_inv = dnnl::memory(vector_transpose_md, engine);

    // 计算 y_true * log(y_pred)
    auto y_true_transpose = dnnl::memory(vector_transpose_md, engine);
    auto identity_desc = dnnl::eltwise_forward::desc(
            prop_kind::forward_training,
            algorithm::eltwise_linear,
            vector_md,
            1.f,
            0.f);
    auto identity_pd = eltwise_forward::primitive_desc(identity_desc, engine);

    auto add1 = dnnl::memory(scalar_md, engine);
    auto ip_desc = inner_product_forward::desc(
            prop_kind::forward_training,
            vector_transpose_md,
            vector_transpose_md,
            scalar_md);

    auto ip_pd = inner_product_forward::primitive_desc(ip_desc, engine);

    // 计算(1-y_true) * log(1-y_pred)
    auto add2 = dnnl::memory(scalar_md, engine);

    // 计算所有值的和
    auto loss_sum = dnnl::memory(scalar_md, engine);

    std::vector<memory::desc> add_vector_md = {scalar_md, scalar_md};
    std::vector<dnnl::memory> add_vector = {add1, add2};
    std::vector<float> scales = {1.f, 1.f};

    auto loss_sum_pd = sum::primitive_desc(scales, add_vector_md, engine);

    // 使用元素线性除以 -n
    auto loss = dnnl::memory(scalar_md, engine);
    float batch_size_inv = (float) 1 / dim_nc[0];
    auto divide_desc = eltwise_forward::desc(
            prop_kind::forward_training,
            algorithm::eltwise_linear,
            scalar_md,
            -batch_size_inv,
            0.f);
    auto divide_pd = dnnl::eltwise_forward::primitive_desc(divide_desc, engine);
}
```

# 致谢
[oneDNN Documentation](https://oneapi-src.github.io/oneDNN/index.html)
[Intel® oneAPI DPC++/C++ Compiler](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler.html#gs.00pus8)
